### ISSUE TITLE
Impossible/abnormal stack

### ISSUE THREAD

**Issue by egwepas** [jvm-profiling-tools/async-profiler#349]
Hi,

I see "impossible" or "abnormal" stacks appear, mainly during long duration profilings (hours).

Symptoms :
 * I see "deep" methods that appears at the root of the call tree.
 * I see a high number of calls in very fast methods (getters, log level testing). To me it looks like the call count of some "neightboor" method ripped to the next one. I'm not sure, but I think it happens after some times, not right from the start.

Usage :
 * I'm still profiling java batches and some have a long duration. I'm starting the profiler using the java library and dumping (stop/resume) the collapsed format periodically in a timer thread.

I know this will be hard to track but I encounter an issue, and maybe you would have an idea ? Maybe a higher risk of a currupted buffer when the result of the profiling gets huge (over 300MB) ?

**Comment by apangin** [comment]
Do you have an example you can share? For me to understand how abnormal stacks look like.

Do these truncated stacks end with `[unknown_Java]`?

Do you use `-XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints` options?

A few notes:

- The profile shows the number of *samples*, not the number of calls of the particular method. Fast methods may be hot, too. They may occur in a large number of samples - this itself is not a sign of a problem. But I can't say for sure without seeing the complete picture.

- Current async-profiler implementation is not tuned for long profiling sessions. The profile has a hard limit of 65535 different stack traces. I got rid of this limitation in [v2.0](https://github.com/jvm-profiling-tools/async-profiler/tree/v2.0) which is currently under development. Meanwhile I can suggest to use `start/stop` instead of `resume/stop` to reset profiling data between sessions. It is possible to [merge](https://github.com/jvm-profiling-tools/async-profiler/issues/333) multiple sessions into a single graph anyway.

  However, even when the stack trace limit is hit, profiler just stops collecting new stack traces, but it still should not result in invalid stacks.

**Comment by egwepas** [comment]
I'm not sure I can share the results as-is but here is what happens :

The normal stack would be :
`[main tid=xxx]packageA/ClassA.main;packageB/ClassB.methodB;packageC/ClassC.methodC;packageD/ClassD.methodD 125`
And sometimes among the legit stacks I see :
`[main tid=xxx]packageC/ClassC.methodC;packageE/ClassE.methodE 2`

I'm describing it as an abnormal stack because this can not happen. The root method nevers call that method directly, it should for example always directly go throught `packageA/ClassA.main` method.

I did not use `-XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints`, i'll try that.

As for your notes
 * I understand the concept of sampling.
 * My biggest profiling result (collapsed format) has 53k lines and could get down to 44k if I were to filter on the main thread. Looks like i'm getting close to the limit, but not reaching it yet.

**Comment by apangin** [comment]
When I investigated another report about truncated stack traces, I found that those traces were returned by the JVM AsyncGetCallTrace function. This was a JVM bug [JDK-8153167](https://bugs.openjdk.java.net/browse/JDK-8153167). I suppose you are observing the same issue.

I'm afraid async-profiler can hardly do anything about it, since it has no way to distinguish correct stack traces returned by the JVM from the truncated stack traces. Fortunately, the amount of such stack traces was relatively small (<1%) so they could be just ignored. What about your case? How many abnormal stack traces is there?
